File: server/src/main/java/io/microconfig/server/configs/ConfigGeneratorImpl.java
package io.microconfig.server.configs;

import io.microconfig.core.Microconfig;
import io.microconfig.core.configtypes.ConfigTypeFilter;
import io.microconfig.core.configtypes.ConfigTypeFilters;
import io.microconfig.core.environments.repository.EnvironmentException;
import io.microconfig.core.properties.ResolveException;
import io.microconfig.core.properties.serializers.ConfigResult;
import io.microconfig.server.git.GitService;
import io.microconfig.server.rest.exceptions.BadRequestException;
import io.microconfig.server.vault.VaultKVSecretResolverStrategy;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.io.File;
import java.util.List;

import static io.microconfig.core.Microconfig.searchConfigsIn;
import static io.microconfig.core.configtypes.ConfigTypeFilters.configTypeWithName;
import static io.microconfig.core.configtypes.ConfigTypeFilters.eachConfigType;
import static io.microconfig.core.properties.serializers.PropertySerializers.asConfigResult;
import static io.microconfig.core.properties.templates.TemplatesService.resolveTemplatesBy;
import static java.util.Collections.emptyList;

@Slf4j
@Service
@RequiredArgsConstructor
public class ConfigGeneratorImpl implements ConfigGenerator {
    private final GitService gitService;

    @Override
    public List<ConfigResult> generateConfigs(String component, String env, ConfigOptions options) {
        try {
            log.info("Generating configs for {} in {} env", component, env);
            var microconfig = initMicroconfig(gitService, options);
            var configType = configTypes(options);
            return microconfig.inEnvironment(env)
                    .findComponentsFrom(emptyList(), List.of(component))
                    .getPropertiesFor(configType)
                    .resolveBy(microconfig.resolver())
                    .forEachComponent(resolveTemplatesBy(microconfig.resolver()))
                    .save(asConfigResult());
        } catch (IllegalArgumentException | EnvironmentException | ResolveException e) {
            throw new BadRequestException(e.getMessage());
        }
    }

    private Microconfig initMicroconfig(GitService gitService, ConfigOptions options) {
        var microconfig = searchConfigsIn(configDir(gitService, options));
        microconfig.logger(false);
        return withAdditionalPlaceholderResolvers(microconfig, options);
    }

    private File configDir(GitService gitService, ConfigOptions options) {
        if (options.tag != null) return gitService.checkoutTag(options.tag);
        if (options.branch != null) return gitService.checkoutBranch(options.branch);
        return gitService.checkoutDefault();
    }

    private Microconfig withAdditionalPlaceholderResolvers(Microconfig microconfig, ConfigOptions options) {
        var dynamicVars = new DynamicVarsResolverStrategy(options.vars);
        var vault = new VaultKVSecretResolverStrategy(microconfig, dynamicVars);
        return microconfig.withAdditionalPlaceholderResolvers(List.of(dynamicVars, vault));
    }

    private static ConfigTypeFilter configTypes(ConfigOptions options) {
        return options.type == null ? eachConfigType() : configTypeWithName(options.getType());
    }
}

File: server/src/main/kotlin/io/microconfig/server/configs/ConfigGeneratorImpl.kt
package io.microconfig.server.configs

import io.microconfig.core.Microconfig
import io.microconfig.core.Microconfig.searchConfigsIn
import io.microconfig.core.configtypes.ConfigTypeFilter
import io.microconfig.core.configtypes.ConfigTypeFilters.configTypeWithName
import io.microconfig.core.configtypes.ConfigTypeFilters.eachConfigType
import io.microconfig.core.properties.Properties
import io.microconfig.core.properties.serializers.PropertySerializers.asConfigResult
import io.microconfig.core.properties.templates.TemplatesService.resolveTemplatesBy
import io.microconfig.server.client.dto.ServiceConfigMap
import io.microconfig.server.client.dto.ServiceConfigRaw
import io.microconfig.server.common.logger
import io.microconfig.server.git.GitService
import io.microconfig.server.vault.VaultKVSecretResolverStrategy
import org.springframework.stereotype.Service

@Service
class ConfigGeneratorImpl(val gitService: GitService) : ConfigGenerator {
    private val log = logger()

    override fun generateConfigs(component: String, env: String, options: ConfigOptions): List<ServiceConfigRaw> {
        return generateProperties(component, env, options)
            .save(asConfigResult())
            .filter { it.content.isNotEmpty() }
            .map { toConfigRaw(it) }
    }

    override fun generateConfigMaps(component: String, env: String, options: ConfigOptions): List<ServiceConfigMap> {
        return generateProperties(component, env, options)
            .save(asConfigMap())
            .filter { it.content.isNotEmpty() }
    }

    private fun generateProperties(component: String, env: String, options: ConfigOptions): Properties {
        log.info("Generating configs for {} in {} env", component, env)
        val microconfig = initMicroconfig(options)
        val configType = configTypes(options)
        return microconfig.inEnvironment(env)
            .findComponentWithName(component)
            .getPropertiesFor(configType)
            .resolveBy(microconfig.resolver())
            .forEachComponent(resolveTemplatesBy(microconfig.resolver()))
    }

    private fun initMicroconfig(options: ConfigOptions): Microconfig {
        val configDir = gitService.checkoutRef(options.ref)
        val microconfig = searchConfigsIn(configDir)
        microconfig.logger(false)
        return microconfig
            .addDynamicVars(options)
            .addVaultResolver()
    }

    private fun Microconfig.addDynamicVars(options: ConfigOptions): Microconfig {
        val dynamicVars = DynamicVarsPropertiesRepository(options.vars)
        return this.withAdditionalPropertiesRepositories(listOf(dynamicVars))
    }

    private fun Microconfig.addVaultResolver(): Microconfig {
        val vault = VaultKVSecretResolverStrategy(this)
        val resolvers = this.additionalPlaceholderResolvers() + vault
        return this.withAdditionalPlaceholderResolvers(resolvers)
    }

    private fun configTypes(options: ConfigOptions): ConfigTypeFilter {
        return if (options.type == null) eachConfigType() else configTypeWithName(options.type)
    }

}